---
title: "NIFA FullCyc2 incorporators workup"
author: "Cassandra Wattenburger"
date: "3/15/2023"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(results = "show")
knitr::opts_chunk$set(message = FALSE)
```

```{r}
# Clear working directory, load in packages, generate package info
rm(list=ls())

library("tidyverse")

sessionInfo()
```

# Import data

```{r}
# Incorporator data from DulLC
incorp <- readRDS("../cluster_fullcyc2/fullcyc2_l2fc_testoutput.rds")
incorp

# Taxonomy of incorporators
incorp_tax <- incorp %>% select(OTU, Phylum:Species)
```

# Reformat data

```{r}
# Reformat and exclude non-significant results
incorp_sig <- incorp %>% 
  mutate(ecosystem = gsub(".substrate == '12C-Con' & day == '[0-9]+' & ecosystem == '([a-z]+)'.+", "\\1", .id), # separate out .id column
         day = gsub(".substrate == '12C-Con' & day == '([0-9]+)' .+", "\\1", .id),
         day = as.numeric(day),
         substrate = gsub(".+substrate == '13C-([a-z]+)'.+", "\\1", .id, ignore.case=T),
         substrate = case_when(substrate=="Pal" ~ "palmitic acid",
                               substrate=="Ami" ~ "amino acid",
                               substrate=="Xyl" ~ "xylose",
                               substrate=="Cel" ~ "cellulose",
                               substrate=="Van" ~ "vanillin")) %>%
  select(OTU, ecosystem, day, substrate, log2FoldChange, padj) %>% 
  filter(padj < 0.05) # filter incorporators, Sam says deseq2 was set up so negative log2foldchange indicates incorporation
  
incorp_sig
```

Create list of incorporator OTUs:

```{r}
incorp_otus <- select(incorp_sig, OTU) %>% 
  unique()
```

```{r, eval=FALSE}
write_tsv(incorp_otus, "../mothur_cluster/incorp_otus.txt")
```

Create list of growing taxa from NIFA experiment:

```{r}
growth_asvs <- readRDS("../data_intermediate/NIFA_grests_final_ind.rds") %>% 
  select(ASV) %>% 
  unique()
```

```{r, eval=FALSE}
write_tsv(growth_asvs, "../mothur_cluster/growth_asvs.txt")
```

# BLAST processing

Import: 

```{r}
blast <- read_tsv("../blast_fullcyc2/nifa_fullcyc2_blastn_parsed.txt")
head(blast)
```

Choose best hit for each ASV (with at least 97% identity):

```{r}
# Select best hit(s)
blast_best <- filter(blast, Identity >= 97) %>% # remove hits lower than 97% identity
  group_by(Query) %>% 
  summarize(Identity = max(Identity)) %>% # choose highest identity match
  ungroup()

blast_best <- semi_join(blast, blast_best)
blast_best

# Tidy up
blast_best <- select(blast_best, ASV=Query, blast_OTU=Hit, blast_identity=Identity, blast_evalue=Evalue)
  
# Some summary
length(unique(blast_best$Query)) # ASVs matched with at least 97% identity
min(blast_best$Identity) # lowest identity match in queries
```

Identify ties:

```{r}
# Isolate ASVs with multiple highest hits
blast_dups_logic <- duplicated(blast_best$ASV)

blast_dups_asvs <- blast_best %>% 
  add_column(duplicated = blast_dups_logic) %>% 
  filter(duplicated==TRUE) %>% 
  select(ASV) %>% 
  unique() %>% 
  as.list()

# How many ASVs with duplicate equally best hits?
length(blast_dups_asvs$ASV)

# Isolate duplicate hits
blast_dups <- blast_best %>% 
  filter(ASV %in% blast_dups_asvs$ASV)

# Isolate single hits
blast_sing <- blast_best %>% 
  filter(!(ASV %in% blast_dups_asvs$ASV))
```

## Deal with multiple OTU hits

For ASVs with two or more OTU hits with equal percentage identity, I'll choose the lowest e-value. If e values also tie, I'll keep the OTU properties of both (or all) OTU hits.

https://www.ncbi.nlm.nih.gov/books/NBK62051/

```{r}
blast_dups_eval <- blast_dups %>% 
  group_by(ASV) %>% 
  summarize(blast_evalue = min(blast_evalue)) %>% # isolate lowest e-value hit
  ungroup()

# Filter by min. e-value
blast_dups <- blast_dups %>% 
  semi_join(blast_dups_eval)

# Remaining duplicates
blast_dups_logic2 <- duplicated(blast_dups$ASV)
blast_dups_asvs2 <- blast_dups %>% 
  add_column(duplicated = blast_dups_logic2) %>% 
  filter(duplicated==TRUE) %>% 
  select(ASV) %>% 
  unique() %>% 
  as.list()

blast_dups2 <- filter(blast_dups, ASV %in% blast_dups_asvs2$ASV)
blast_dups2
```

Look at taxonomy of duplicate OTU hits:

```{r}
# Isolate OTU hits from taxonomy info
tax_dups <- filter(incorp_tax, OTU %in% unique(blast_dups2$blast_OTU)) %>% 
  unique()

# Merge with blast results
blast_dups2 <- blast_dups2 %>% 
  left_join(tax_dups, by=join_by(blast_OTU==OTU))

blast_dups2
```

Visualize:

```{r}
# Phylum level
blast_dups2 %>% 
  ggplot(aes(x=Phylum, y=blast_identity, color=ASV)) +
  geom_point(alpha=0.8) +
  facet_wrap(~ASV, scales = "free") +
  theme_test() +
  theme(legend.position = "none",
        axis.text.y = element_blank())

# Class
blast_dups2 %>% 
  ggplot(aes(x=Class, y=blast_identity, color=ASV)) +
  geom_point(alpha=0.8) +
  facet_wrap(~ASV, scales = "free") +
  theme_test() +
  theme(legend.position = "none",
        axis.text.y = element_blank())

# Order
blast_dups2 %>% 
  ggplot(aes(x=Order, y=blast_identity, color=ASV)) +
  geom_point(alpha=0.8) +
  facet_wrap(~ASV, scales = "free") +
  theme_test() +
  theme(legend.position = "none",
        axis.text.y = element_blank())

# Family
blast_dups2 %>% 
  ggplot(aes(x=Family, y=blast_identity, color=ASV)) +
  geom_point(alpha=0.8) +
  facet_wrap(~ASV, scales = "free") +
  theme_test() +
  theme(legend.position = "none",
        axis.text.y = element_blank())

# Genus
blast_dups2 %>% 
  ggplot(aes(x=Genus, y=blast_identity, color=ASV)) +
  geom_point(alpha=0.8) +
  facet_wrap(~ASV, scales = "free") +
  theme_test() +
  theme(legend.position = "none",
        axis.text.y = element_blank())

# Species
blast_dups2 %>% 
  ggplot(aes(x=Species, y=blast_identity, color=ASV)) +
  geom_point(alpha=0.8) +
  facet_wrap(~ASV, scales = "free") +
  theme_test() +
  theme(legend.position = "none",
        axis.text.y = element_blank())
```

These plots tell me if the duplicate OTU hits fall in the same or different taxonomic groups. One dot per graph means they all are within the same group.

Duplicate OTU hits start diverging ~ order level, but the splits are among "uncultured", "ambigious", or "NA" groups which tells me these are uncultured and uncharacterized which would explain variation in higher resolution taxonomic placement. Overall taxa are very closely related, as expected. Likely OK to merge the substrate use qualities for the ASVs

Combine single and duplicate blast results:

```{r}
# Put single hits back in with duplicate hits
blast_mh <- bind_rows(blast_sing, blast_dups) %>% 
  mutate(multiple_hit=if_else(ASV %in% blast_dups_asvs2$ASV, "yes", "no")) # Add multiple hit warning column

blast_mh
```

## Complete incorporation profiles for each matched OTU

```{r}
# # Reformat incorporator data
# incorp_sig <- incorp_sig %>% 
#   mutate(label=paste0(OTU, ecosystem, day)) # create id column
# 
# incorp_meta <- select(incorp_sig, label, OTU, ecosystem, day) %>%
#   unique()
# 
# # Complete variables for each incorporation observation yes/no
# incorp_yn <- incorp_sig %>% 
#   select(label, OTU, ecosystem, day, substrate)
#   
# 
# incorp_yn

# ARCHIVE
# Complete variables for each incorporation observation yes/no
# incorp_yn <- incorp_sig %>% 
#   add_column(incorporator = "yes") %>% # adding this because I'll need to label non-incorporators later too
#   select(label, OTU, ecosystem, day, substrate, incorporator) %>% 
#   pivot_wider(id_cols=label, names_from=substrate, values_from=incorporator) %>% # label non-incorporation for each OTU
#   mutate_all(~ if_else(is.na(.x), "no", .x)) %>%
#   rename(palmitic_acid='palmitic acid', amino_acid='amino acid') %>% 
#   left_join(incorp_meta) %>% 
#   select(label, OTU, ecosystem, day, palmitic_acid:vanillin)
# 
# incorp_yn
```

## Simplify ecosystems

Create union of substrate incorporation properties among different ecosystems for the same OTU, was the OTU labelled and if so, what was the earliest day it was labelled in any ecosystem?

```{r}
# Simplify
incorp_simp <- data.frame()
for (o in unique(incorp_sig$OTU)) {
  for (s in unique(incorp_sig[incorp_sig$OTU==o,]$substrate)) {
    data_os <- filter(incorp_sig, OTU==o & substrate==s) # isolate substrates that were incorporated
    first_day <- min(data_os$day)
    this_row <- bind_cols(OTU=o, substrate=s, first_day=first_day)
    incorp_simp <- bind_rows(incorp_simp, this_row)
  }
}

incorp_simp
```

## Join with BLAST results

```{r}
blast_incorp <- full_join(incorp_simp, blast_mh, by=join_by(OTU==blast_OTU)) %>% 
  filter(!is.na(ASV)) %>% # remove incorporator OTUs that did not match an ASV
  rename(blast_OTU=OTU, incorp_firstday=first_day)

blast_incorp
```

## Simplify mutliple hits

Create union of incorporation profiles for OTUs that equally matched one ASV.

```{r}
# Parse
blast_incorp_simp <- data.frame()
for (a in unique(blast_incorp$ASV)) {
  data_a <- filter(blast_incorp, ASV==a) %>% 
    group_by(ASV, substrate, blast_identity, blast_evalue) %>% 
    summarize(incorp_firstday = min(incorp_firstday)) %>%  # isolate first day of incorporation among multiple hits (if any)
    ungroup() %>% 
    unique()
  blast_incorp_simp <- bind_rows(blast_incorp_simp, data_a)
}

# Visualize
blast_incorp_simp %>% 
  ggplot(aes(x=substrate, y=incorp_firstday)) +
  geom_jitter() +
  geom_hline(yintercept=c(3,6,14,30), linetype=2) +
  theme_test()
```

There is variation in first day of incorporation across substrates, so variable is worth keeping around.

# Summarize incorporator hits to ASVs

BLAST % identity:

```{r}
mean(blast_incorp_simp$blast_identity)
median(blast_incorp_simp$blast_identity)
min(blast_incorp_simp$blast_identity)
max(blast_incorp_simp$blast_identity)
hist(blast_incorp_simp$blast_identity, main="Histogram", xlab="Percent identity to nearest BLAST hit")
```

BLAST e-values:

```{r}
mean(blast_incorp_simp$blast_evalue)
median(blast_incorp_simp$blast_evalue)
min(blast_incorp_simp$blast_evalue)
max(blast_incorp_simp$blast_evalue)
hist(blast_incorp_simp$blast_evalue, main="Histogram", xlab="Evalue of nearest BLAST hit")
```

Number of incorporation hits per substrate:

```{r}
# Summary table
blast_incorp_simp %>% 
  group_by(substrate) %>% 
  summarize(total = n())

# Visualize
blast_incorp_simp %>% 
  group_by(substrate) %>% 
  summarize(total = n()) %>% 
  ungroup() %>%
  ggplot(aes(x=reorder(substrate, -total), y=total)) +
  geom_col() +
  labs(title="ASV hits to incorporators", y="ASVs capable of incorporation", x="substrate") +
  theme_test()

blast_incorp_simp %>% 
  group_by(substrate, incorp_firstday) %>% 
  summarize(total = n()) %>% 
  ungroup() %>%
  ggplot(aes(x=reorder(substrate, -total), y=total)) +
  geom_col() +
  facet_wrap(~incorp_firstday) +
  labs(title="ASV hits to incorporators by day of earliest incorporation observed", y="ASVs capable of incorporation", x="substrate") +
  theme_test() +
  theme(axis.text.x=element_text(angle=45, hjust=1))
```

# Exploration

## Incorporation by soil
# BOOKMARK
```{r}

```


### Incorporation vs growth

Add growth metrics and genomic characteristics based on ASVs:

```{r}
# Paprica estimates
pap <- readRDS("../data_intermediate/NIFA_paprica_ests.rds") %>% 
  select(ASV, n16S)

# Growth estimates
growth_asv <- readRDS("../data_intermediate/NIFA_grests_final_ind.rds") %>% 
  mutate(length = end_day-start_day) %>% 
  select(ASV, Soil, Replicate, g, start_day, length, start_abund, end_abund, change_abund) %>% 
  left_join(pap) %>% 
  group_by(ASV, Soil) %>% 
  summarize(g = mean(g),
            start_day = mean(start_day),
            length = mean(length),
            start_abund_corr = mean(start_abund)/mean(n16S), 
            end_abund_corr = mean(end_abund)/mean(n16S),
            change_abund_corr = mean(change_abund)/mean(n16S),
            n16S = mean(n16S)) %>% 
  ungroup()

# Combine with incorporation profiles
growth_incorp <- growth_asv %>% 
  right_join(blast_incorp_simp)

growth_incorp
```

Visualize:

```{r}
# g
growth_incorp %>% 
  ggplot(aes(x=substrate, y=log(g), color=Soil)) +
  geom_violin() +
  geom_jitter(alpha=0.5) +
  theme_test()

# lag
growth_incorp %>% 
  ggplot(aes(x=substrate, y=start_day, color=Soil)) +
  geom_violin() +
  geom_jitter(alpha=0.5) +
  theme_test()

# Length
growth_incorp %>% 
  ggplot(aes(x=substrate, y=length, color=Soil)) +
  geom_violin() +
  geom_jitter(alpha=0.5) +
  theme_test()

# start abund
growth_incorp %>% 
  ggplot(aes(x=substrate, y=log(start_abund_corr), color=Soil)) +
  geom_violin() +
  geom_jitter(alpha=0.5) +
  theme_test()

# end abund
growth_incorp %>% 
  ggplot(aes(x=substrate, y=log(end_abund_corr), color=Soil)) +
  geom_violin() +
  geom_jitter(alpha=0.5) +
  theme_test()

# change abund corr
growth_incorp %>% 
  ggplot(aes(x=substrate, y=log(change_abund_corr), color=Soil)) +
  geom_violin() +
  geom_jitter(alpha=0.5) +
  theme_test()

# n16S
growth_incorp %>% 
  ggplot(aes(x=substrate, y=n16S, color=Soil)) +
  geom_violin() +
  geom_jitter(alpha=0.5) +
  theme_test()
```

# Create dataframe for clustering

```{r}
# Reformat

# Wide format
clustering_df <- growth_incorp %>% 
  select(ASV:n16S, substrate, incorp_firstday) %>% 
  mutate(incorp_firstday = as.character(incorp_firstday)) %>%
  pivot_wider(id_cols = c(ASV:n16S), names_from=substrate, values_from=incorp_firstday) %>% # combining substrate incorporation and day of incorp vars
  rename(palmitic_acid='palmitic acid', amino_acid='amino acid') 

# Separate variables for first incorporation day
incorp_firstday <- clustering_df %>% 
  select(cellulose:xylose) %>% 
  rename(cellulose_firstday=cellulose, amino_firstday=amino_acid, palmitic_firstday=palmitic_acid, vanillin_firstday=vanillin, xylose_firstday=xylose) %>% 
  mutate_all(function(x) if_else(is.na(x), "NA", x))

# Put it all together
clustering_df2 <- clustering_df %>% 
  bind_cols(incorp_firstday) %>% 
  mutate(across(c(cellulose, amino_acid, palmitic_acid, vanillin, xylose), function(x) if_else(is.na(x), "no", "yes")), # binary yes/no incorporation
         label = paste0(Soil, "_", ASV)) %>% # unique observation labels for each soil and asv
  select(label, g:xylose_firstday)

clustering_df2
```

Save:

```{r, eval=FALSE}
saveRDS(clustering_df2, file="../data_intermediate/NIFA_cluster_variables.rds")
```

